### ISSUE TITLE
Improve referencing __version__ attribute.

### ISSUE THREAD

**Issue by jenstroeger** [commitizen-tools/commitizen#913]
### Description

The other day I came across the package version declaration in SQLAlchemy’s `setup.cfg` [here](https://github.com/sqlalchemy/sqlalchemy/blob/900d13acb4f19de955eb609dea52a755f0d11acb/setup.cfg#L3):
```
version = attr: sqlalchemy.__version__
```
The handling of special directive `attr:` is documented [here](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html#specifying-values).

I thought this quite useful to ensure that a version is declared in one place only; currently `cz` relies on [version_files](https://commitizen-tools.github.io/commitizen/bump/#version_files) to point at attributes that need to be bumped.

This is probably a low-priority item that may or may not make sense. But I wanted to raise it for discussion anyway…

**Comment by woile** [comment]
What about [custom version providers](https://commitizen-tools.github.io/commitizen/config/#custom-version-provider)?
I think you could have one using `__version__` as the source of the version

**Comment by jenstroeger** [comment]
Hmm. That’s a nice idea but I’d have to wire my own blob of code into into `pyproject.toml` and ship another module, right? Unless we allow folks to stash Python into a TOML file (but that’s getting ugly for little benefit…) 🤔

**Comment by woile** [comment]
You'd have to publish a package with the custom provider in python, no code goes in pyproject.toml.

Then you'd install:

```sh
pip install commitizen cz-attr-version
```

And then you'd do in commitizen toml:

```toml
provider = "attr_version" 
```

**Comment by woile** [comment]
Now that I think about it... maybe we could even make the existing provider `pep621` read this, right?
pep621 is reading from the `version` field, that means it should be able to interpret accepted values, either a version or an attr 🤔 

On the other hand, the recommendation seems to point to dropping the usage of any version information at module leven in favor of just using `importlib.metadata.version()`

[see pep-396-rejection](https://peps.python.org/pep-0396/#pep-rejection)

Thoughts @Lee-W @noirbizarre ?

**Comment by jenstroeger** [comment]
> Now that I think about it... maybe we could even make the existing provider `pep621` read this, right? 

I do like that.

> On the other hand, the recommendation seems to point to dropping the usage of any version information at module leven in favor of just using `importlib.metadata.version()` [see pep-396-rejection](https://peps.python.org/pep-0396/#pep-rejection)

That PEP is from 2011, so I’m tempted to _not_ give it too much weight. Might be worth bringing up for discussion at [the forum](https://discuss.python.org/c/packaging/14)?

**Comment by Lee-W** [comment]
> Now that I think about it... maybe we could even make the existing provider pep621 read this, right?

like this one as well

**Comment by noirbizarre** [comment]
This is a complex topic I know quite well.
You need to include in the reasoning that while PEP621 is about exposing static metadata that can be read at packaging time (which is perfect for commitizen as it only imply reading a toml file), both `importlib.metadata.version()` and `<root_module>.__version__` are runtime way of accessing the version, and I think this is a bit out of scope for commitizen.

Runtime access to both of those is not yet standardized, PEP621 is not about this and I don't think commitizen has any business to to on this. Those are the responsibility:
- of the package manager to properly set [the version as specified in Core Metadata](https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata) at packing time (this is what you get using `importlib.metadata`)
- of the developper to properly expose the version in `<root_module>.__version__` (given this one is not specified at all and is more a convention)

However, both `importlib.metadata.version()` and `<root_module>.__version__` SHOULD reflect PEP621 versioning. Given the lack of PEP on the topic, there are many ways of handling this.

Here's my recipe to have commitizen properly works with PEP621 metadata and having both runtime access properly work (and my PR on version providers was meant to allow that):
- In my package I actually use `pdm` as package manager with the `dynamic = ["version"]` as PEP621 dynamic versioning and `pdm` scm provider (which exist for setuptools, poetry... as they are all forks or copy-paste from the `setuptools-scm` plugin) which leads to:
  - having a generated version file containing either the scm tag, either a version derived from the last tag and last commit has)
  - having the same version being inject in Core Metadata
- I use commitizen SCM version provider (and not the PEP621) to read the version from the scm
- I have a snippet in my `<root_module>.__init__.py` which is trying to fetch the version from `importlib.metadata` and then fallback on reading the generated version file to expose it on `__version__`

This makes `pyproject.toml` looks like this:
```toml
[project]
...
dynamic = ["version"]
dependencies = [
    "...",
    # importlib from Python 3.9+ API changed in a non-backward compatible way
    "importlib-resources>=1.3;python_version<'3.9'",
]
[tool.pdm.version]
source = "scm"
write_to = "<root_module>/VERSION"
```

and the `<root_module>/__init__.py` looks like this:
```python
from __future__ import annotations

from importlib.metadata import PackageNotFoundError, version

try:
    from importlib.resources import files
except ImportError:
    from importlib_resources import files  # type: ignore[no-redef,import-not-found]

def read_version() -> str:
    try:
        return version("<package name>")
    except (PackageNotFoundError, ImportError):
        version_file = files(__package__) / "VERSION"
        return version_file.read_text() if version_file.is_file() else "0.0.0.dev"


__version__ = read_version()
```

Why not just read the version file ? Because SCM tags don't support PEP440 build identifiers (the `+something` part which can be set at package time on some package manager), the version file generated by the scm plugin use the build identifier for non tagged version to add a commitish+hash and the real build identifier is set in Core Metadata so `importlib.metadata` gives the version that was actually packaged, included all dynamic parts especially those that could be set in CI, while the version file give the version generated from the scm.

**Note**: this is very related to how you expect to use those. For many people (especially those coming from the JS world where this is the `package.json` "normal way"), only tagged versions are relevant and so PEP621 version provider is enough.

TLDR: the PEP621 provider should not read those packaging or runtime only information (meaning they exist at a time commitizen can't read) however documenting how it's possible to properly read those metadata at runtime and supports all those convention in a standard Python package could be a great addition.

**Note**: a static tool trying to read the `__version__` field should be considered as broken as it is an unspecified runtime-only facility which might rely on a lot of dynamic behavior and will most of the time break on static access.
